src/main/java/teammates/logic/core/StudentsLogic.java

```diff
@@ -204,10 +205,42 @@ public void resetStudentGoogleId(String originalEmail, String courseId)
                             .withGoogleId(null)
                             .build());
         } catch (InvalidParametersException | EntityAlreadyExistsException e) {
-            Assumption.fail("Resting google ID shall not cause: " + e.getMessage());
+            Assumption.fail("Resetting google ID shall not cause: " + e.getMessage());
         }
     }
 
+    /**
+     * Regenerates the registration key for the student with email address {@code email} in course {@code courseId}.
+     *
+     * @return the student attributes with the new registration key.
+     * @throws RegenerateStudentException if the newly generated course student has the same registration key as the
+     *          original one.
+     * @throws EntityDoesNotExistException if the student does not exist.
+     */
+    public StudentAttributes regenerateStudentRegistrationKey(String courseId, String email)
+            throws EntityDoesNotExistException, RegenerateStudentException {
+
+        try {
+            StudentAttributes originalStudent = studentsDb.getStudentForEmail(courseId, email);
+            if (originalStudent == null) {
+                throw new EntityDoesNotExistException("Student does not exist: [" + courseId + "/" + email + "]");
+            }
+
+            studentsDb.deleteStudent(courseId, email);
+
+            StudentAttributes updatedStudent = studentsDb.createEntity(originalStudent);
+            if (updatedStudent.getKey().equals(originalStudent.getKey())) {
+                throw new RegenerateStudentException("Regenerated student has same registration key as original student.");
```

Comment:
- At this block, it looks like the new student entity has been persisted to the DB. Is it safe?
- If I understand correctly, this happens by chance if the key (which is supposedly RNG-generated) is identical. If that is the case, shouldn't there be retries?

Reply:
Good catch. I've moved the logic to `StudentsDb` and the changes are only persisted after checking that the new key is different.

I've added a limit of 3 retries. I think that is suitable. The exception is thrown if all 3 tries fail.